### ISSUE TITLE
Data resource group resource errors out when used in module against implicit dependency

### ISSUE THREAD

**Issue by cveld** [hashicorp/terraform-provider-azurerm#21615]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

1.4.6

### AzureRM Provider Version

3.54.0

### Affected Resource(s)/Data Source(s)

azurerm_resource_group

### Terraform Configuration Files

main.tf:

```hcl
resource "azurerm_resource_group" "shared" {
  name     = "shared"
  location = "westeurope"
}

module "test" {
  source = "./Module1"
  name   = azurerm_resource_group.shared.name
}
```

./Module1/test/main.tf:

```hcl
data "azurerm_resource_group" "rg" {
  name = var.name
}
```

### Debug Output/Panic Output

```shell
╷
│ Error: Error: Resource Group "shared" was not found
│
│   with module.test.data.azurerm_resource_group.rg,
│   on Module1\main.tf line 1, in data "azurerm_resource_group" "rg":
│    1: data "azurerm_resource_group" "rg" {
│
╵
```

### Expected Behaviour

The module should not be computed during plan phase as the resource group output is not yet available. 

### Actual Behaviour

Apparently the resource group output is already available during plan phase and is fed into the module. The data resource block inside the module doesn't know it has to wait on its implicit dependency and just tries to read an existing resource group, which fails.

### Steps to Reproduce

Run:

`terraform plan`

### Important Factoids

Interestingly when the data resource is residing in the same module as where the resource group is located, then the data resource is not read during plan phase. Only when you put the data resource inside a module, this error is yielded.

**Comment by askingcat** [comment]
JFYI: In the terraform repository itself this issue has been reported multiple times in the last years (see e.g. [#28511](https://github.com/hashicorp/terraform/issues/28511) and [#26074](https://github.com/hashicorp/terraform/issues/26074)). 

So far I did not see a conclusive answer, but [the response in the open issue from 2020](https://github.com/hashicorp/terraform/issues/26074#issuecomment-693696693) might be useful:

> Unfortunately, because Azure uses user-provided names rather than opaque, auto-generated identifiers, `azurerm_resource_group.test.name` has a known value during the plan step and so Terraform believes it should be able to read that data resource. That then fails because the object hasn't been created yet.

**Comment by rcskosir** [comment]
Thanks for taking the time to submit this issue. It looks like this is a duplicate of #26074, thank you @askingcat for helping find that. We like to try to keep discussions consolidated, so we’re going to close this new issue in favor of #26074.
